### ISSUE TITLE
PreBuild event/target runs twice - during Build and Publish

### ISSUE THREAD

**Issue by EatonZ** [dotnet/sdk#3707]
**Visual Studio Version**: 16.3.1

**Summary / Steps to Reproduce / Behavior**:
I have a .NET Core 3.0 WinForms application and I have added a prebuild event to my project. It gets added to the csproj like this:
```
<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
  <Exec Command="$(ProjectDir)PreBuildTasks.exe" />
</Target>
```
It basically launches my prebuild exe, which does a few things, like increment the version in AssemblyInfo. This has always worked fine until I started experimenting with .NET Core and publishing.

For context, I am going to add this right before the `Exec Command`:
`<Message Importance="high" Text="This is a message" />`

When I publish the project (right click project, Publish), this is what I see:
![2019-10-01_1-25-35](https://user-images.githubusercontent.com/3020941/65936781-cac15b80-e3eb-11e9-93cf-d41efb6adc93.png)

It's clear the prebuild task is running twice. I spent a while researching a way to resolve this problem. The closest solution I found was [this](https://developercommunity.visualstudio.com/content/problem/30395/aspnet-publish-process-runs-build-3-times.html). I added `Condition="'$(DeployOnBuild)' != 'true'"` as directed, but unfortunately it did not do anything. The other solution in that thread says the issue is fixed, so what is happening here?

This behavior may be by design, but I decided to open this issue in case it is not, and because I'd like some guidance on the correct way to solve this problem if I am going at this the wrong way. I only want my prebuild task to run on **Build**, not **Publish**.

**Comment by EatonZ** [comment]
@jmarolf @davidwengier Ok, thanks for moving. At first glance it seemed like the project repo was the best place.

**Comment by livarcocc** [comment]
cc @dsplaisted @nguerrera

**Comment by EatonZ** [comment]
Any further thoughts on this?

**Comment by daalcant** [comment]
Also curious to hear if there are any updates on this. Experienced a similar issue with "dotnet publish" running pre-build tasks twice when executed from a directory containing the .sln file, but not when running it from sub-folders containing their own .csproj files.

**Comment by holc-tkomp** [comment]
Blazor WebAssembly App's folder publish seems to call `CoreBuild` target twice as well (probably other pre-build ones as well). Once during Build and once during Publish.
```xml
<PropertyGroup>
  <TargetFramework>net5.0</TargetFramework>
  <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
  <UseBlazorWebAssembly>true</UseBlazorWebAssembly>
</PropertyGroup>
```
`Condition="'$(DeployOnBuild)' != 'true'"` helped here though.

**Comment by EatonZ** [comment]
No change in .NET 5 and VS 16.8. Would love to hear if any consideration is being given to this.

**Comment by dsplaisted** [comment]
I don't think we have any real plans to fix this.  From the command line, the target would probably be called only once if you ran `dotnet publish`.  If it's getting run twice, then investigating a binary log should help reveal the reason why.  From VS, I'm not sure there's any way to stop it from calling the build targets twice.

FYI @vijayrkn @richlander @KathleenDollard

**Comment by EatonZ** [comment]
@dsplaisted @vijayrkn @richlander @KathleenDollard If there is no plan to fix, then can you make the workaround a little better? Such as a new property I suggested in https://github.com/dotnet/sdk/issues/3707#issuecomment-537161696.

**Comment by unix786** [comment]
`dotnet publish` from command line (power shell) works fine, but if I have that particular project open in Visual Studio, targets are run twice, even though I'm still running publish from the command line. But if I unload the project, target duplication stops (I can leave VS running).

This seems to be caused by the fact that some of the referenced projects have multiple `TargetFrameworks` ([as was pointed out](https://github.com/dotnet/sdk/issues/3707#issuecomment-642235692)), even though the **top project targets only a single framework**. [This post at stackoverflow.com](https://stackoverflow.com/a/46679661/5050045) mentions `IsCrossTargetingBuild`, which is defined in [Microsoft.Managed.Before.targets](https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.Managed.Before.targets), but that one does not work when the top project has only one target.

Some side notes:
- Repeated runs don't produce any messages in "build" output or the power shell console.
- `dotnet build` seems to produce even more repeated runs than `dotnet publish`
- Checking `PublishProtocol` did not help. It desn't seem to get defined in my case.

### The Workaround
When the top project has only one target, all runs get the same `TargetFramework`, but when calling publish and specifying **extra parameters in the command line**, e.g. `dotnet publish -p:MyPublishFlag=True`, they **get passed only to the primary run**. This "primary" run is not necessarily the first one. I use this workaround in conjunction with a target after `CompileTypeScriptWithTSConfig`.

**Comment by Symbai** [comment]
One of our applications is obfuscated during build. For a single publish this process does take not less than 10 minutes. Publishing through VS run this process twice so for each build it takes more than 20 minutes to finish. I'm glad this has been noticed and reported already but sad to see that after more than 2 years it's still not fixed.

**Comment by zrsio4** [comment]
I found this to be useful in my .NET 5 project:
`<Target Name="PreBuild" BeforeTargets="PreBuildEvent" Condition="'$(DeployOnBuild)' != 'true'">`
Note the use of $(DeployOnBuild). This value is empty on build, and true on publish. Now my build events only run during build!
Found here: [https://developercommunity.visualstudio.com/t/postbuildevent-runs-twice-during-publish-build-pub/27585](https://developercommunity.visualstudio.com/t/postbuildevent-runs-twice-during-publish-build-pub/27585)

**Comment by EatonZ** [comment]
Let's keep this open, would be nice to be able to clean up my build process one day to remove the workaround for this problem. 🙂
